actioncable/lib/action_cable/connection/web_socket.rb

```diff
@@ -34,8 +32,21 @@ def rack_response
         websocket.rack_response
       end
 
+      # by order of preference, the first supported client will be chosen
+      CLIENT_SOCKET_CANDIDATES = [ClientRackSocket, ClientFayeSocket]
+
+      def self.establish_connection(env, event_target, event_loop, protocols)
+        @client_socket_klass ||= client_socket_selector(env)
```

Comment:
Memoizing the socket class feels like a stand-in for explicit app configuration. We could clearly state whether we're using a server than provides WebSocket support via `rack.upgrade?` or not.

Reply:
Although your are absolutely right, I think we should keep it this way.

At the moment server don't publish the fact that they support `rack.upgrade?` until the first valid WebSocket request arrives and not many developers are aware of this possible optimization.

This means two things: (1) it's impossible to detect effectively during initialization and (2) it's unlikely that developers will know to set (or un-set) this option.

Moreover, making this into an explicit setting makes application servers harder to change. I can imagine someone moving from iodine to Thin and forgetting to set the option back... especially when many developers are involved.

On the other hand, this is a simple solution with a minimal performance cost that occurs only once.

Reply:
It is simple, it is low cost, yet it's also hidden behavior that's hard to understand or predict.

A clearer alternative would be requiring `rack.upgrade?` support in order to run Action Cable at all, along with providing a middleware that emulates `rack.upgrade?` support for Rack handlers that don't yet provide it natively.

Reply:
That's true.

The memoization does introduce hidden behavior.

I also agree that extracting the network logic from Rails into an optional middleware is probably a positive eventual outcome... however, I think it's too early.

*The memoization* could be safely removed considering the fact the the performance benefit is a single conditional (`env['rack.upgrade?'] == :websocket`). IMHO, this will eliminate the most significant hidden behavior. i.e.,:

```ruby
(env.has_key?('rack.upgrade?') ? ClientRackSocket : ClientFayeSocket).attempt(env, event_target, event_loop, protocols)
```

The immediate extraction of the network logic will delay Rails development and prove costly in developer hours, since it prioritizes the authoring of the network layer in middleware form.

IMHO it would be better to do this work in stages:

1. First allow the `env['rack.upgrade?'] ` to be slowly adopted by the major application servers (Puma, Passenger, etc'), marking this PR as the beginning of a transitional period.

   This will minimize the commitment performed by Rails until the `rack.upgrade?` API becomes a de-facto standard for the major servers. 

2. Extracting the network logic from Rails and (possibly) authoring an optional Rack middleware in the Rack repo.

    Since the major servers will support the `rack.upgrade?` API by this stage, than the authoring of the middleware doesn't need to be prioritized.

This approach was also what I understood from @tenderlove 's [comments during the `rack.upgrade?` API discussion](https://github.com/rack/rack/pull/1272#issuecomment-389285061).

Reply:
Was I off the mark?